url1a41b8e9f2_mr1azl_INF345_tgz-pJ8-TestParserJ8/src/xx/MonCompiler.java

```diff
@@ -0,0 +1,2234 @@
+						&& nodeList.item(i) instanceof Comment) {
+					ret.add(nodeList.item(i));
+				}
+			}
+			if (nodeTest instanceof NodeTypeTest) {
+				if (nodeTest.getNodeType() == NODE_TYPE_PI
+						&& nodeList.item(i) instanceof ProcessingInstruction) {
+					ProcessingInstruction test = (ProcessingInstruction) nodeList
+							.item(i);
+					if (test.getTarget().equals(
+							((NodeTest) nodeTest).getNodeValue())
+							|| ((NodeTest) nodeTest).getNodeValue() == null) {
+						ret.add(nodeList.item(i));
+					}
+				}
+			}
+			if (nodeTest instanceof NodeTypeTest) {
+				if (nodeTest.getNodeType() == NODE_TYPE_NODE
+						&& nodeList.item(i) instanceof Node) {
+					ret.add(nodeList.item(i));
+				}
+			}
+		}
+		return ret;
+	}
+/**
+ * 
+ * @param node
+ * @param ret
+ * @param nodeTest
+ * @param ignore
+ */
+	private static void addParents(Node node, List<Node> ret,
+			NodeTest nodeTest, boolean ignore) {
+		Node parent = node.getParentNode();
+		if (parent != null && !ret.contains(parent)) {
+			addParents(parent, ret, nodeTest, ignore);
+			if (testNode(nodeTest, parent) || ignore) {
+				ret.add(parent);
+			}
+		}
+
+	}
+/**
+ * 
+ * @param node
+ * @param ret
+ * @param nodeTest
+ */
+	private static void addDescendant(Node node, List<Node> ret,
+			NodeTest nodeTest) {
+		for (int i = 0; i < node.getChildNodes().getLength(); i++) {
+			NodeList nodeList = node.getChildNodes();
+			if (!(nodeList.item(i) instanceof Comment
+					|| nodeList.item(i) instanceof ProcessingInstruction || nodeList
+						.item(i) instanceof Text)) {
+				if (testNode(nodeTest, node.getChildNodes().item(i))) {
+					ret.add(node.getChildNodes().item(i));
+				}
+				addDescendant(node.getChildNodes().item(i), ret, nodeTest);
+			}
+
+		}
+	}
+
+	/**
+	 * 
+	 * @param test
+	 * @param node
+	
+	 * @return boolean
+	 */
+	public static boolean testNode(NodeTest test, Node node) {
+		if (test == null) {
+			return true;
+		}
+		if (test instanceof NodeNameTest) {
+
+			NodeNameTest nodeNameTest = (NodeNameTest) test;
+			QName testName = nodeNameTest.getNodeQName();
+			QName nodeName = new QName(node.getNamespaceURI(),
+					node.getNodeName());
+			if (nodeName == null) {
+				return false;
+			}
+
+			String testPrefix = testName.getPrefix();
+			String nodePrefix = nodeName.getPrefix();
+			if (!equalStrings(testPrefix, nodePrefix)) {
+				String testNS = nodeNameTest.getNamespaceURI();
+				String nodeNS = node.getNamespaceURI();
+				if (!equalStrings(testNS, nodeNS)) {
+					return false;
+				}
+			}
+			if (nodeNameTest.isWildcard()) {
+				return true;
+			}
+			return testName.getLocalPart().equals(nodeName.getLocalPart());
+		}
+		return test instanceof NodeTypeTest
+				&& ((NodeTypeTest) test).getNodeType() == Compiler.NODE_TYPE_NODE;
+	}
+/**
+ * 
+ * @param s1
+ * @param s2
+
+ * @return boolean
+ */
+	private static boolean equalStrings(String s1, String s2) {
+		return s1 == s2 || s1 != null && s1.equals(s2);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An equality operator (== or !=) is used to compare strings, which matches strings based on address.

**Fix**: Do content comparison using the equals() method to compare the values of the strings.